massa-protocol-worker/src/worker_operations_impl.rs

```diff
@@ -0,0 +1,221 @@
+            let _ = self.ban_node(&node_id).await;
+        }
+    }
+
+    /// Clear the `asked_operations` data structure and reset
+    /// `ask_operations_timer`
+    pub(crate) fn prune_asked_operations(
+        &mut self,
+        ask_operations_timer: &mut std::pin::Pin<&mut Sleep>,
+    ) -> Result<(), ProtocolError> {
+        self.asked_operations.clear();
+        // reset timer
+        let instant = Instant::now()
+            .checked_add(
+                self.protocol_settings
+                    .asked_operations_pruning_period
+                    .into(),
+            )
+            .ok_or(TimeError::TimeOverflowError)?;
+        ask_operations_timer.set(sleep_until(instant));
+        Ok(())
+    }
+
+    pub(crate) async fn update_ask_operation(
+        &mut self,
+        operation_batch_proc_period_timer: &mut std::pin::Pin<&mut Sleep>,
+    ) -> Result<(), ProtocolError> {
+        let now = Instant::now();
+        while !self.op_batch_buffer.is_empty()
+        // This unwrap is ok because we checked that it's not empty just before.
+            && now >= self.op_batch_buffer.front().unwrap().instant
+        {
+            let op_batch_item = self.op_batch_buffer.pop_front().unwrap();
+            debug!("depile");
+            self.on_batch_operations_received(op_batch_item.operations_ids, op_batch_item.node_id)
+                .await?;
+        }
+        // reset timer
+        if let Some(item) = self.op_batch_buffer.front() {
+            operation_batch_proc_period_timer.set(sleep_until(item.instant));
+        } else {
+            let next_tick = now
+                .checked_add(self.protocol_settings.operation_batch_proc_period.into())
+                .ok_or(TimeError::TimeOverflowError)?;
+            operation_batch_proc_period_timer.set(sleep_until(next_tick));
+        }
+        Ok(())
+    }
+
+    /// Process the reception of a batch of asked operations, that means that
+    /// we sent already a batch of ids in the network notifying that we already
+    /// have those operations. Ask pool for the operations (will change with
+    /// the shared storage)
+    ///
+    /// See also `on_operation_results_from_pool`
+    pub(crate) async fn on_asked_operations_received(
+        &mut self,
+        node_id: NodeId,
+        op_ids: OperationIds,
+    ) -> Result<(), ProtocolError> {
+        if let Some(node_info) = self.active_nodes.get_mut(&node_id) {
+            for op_ids in op_ids.iter() {
+                node_info.known_operations.remove(op_ids);
+            }
+        }
+        let mut operation_ids = OperationIds::default();
+        for op_id in op_ids.iter() {
+            if self.checked_operations.get(op_id).is_some() {
+                operation_ids.insert(*op_id);
+            }
+        }
+        self.send_protocol_pool_event(ProtocolPoolEvent::GetOperations((node_id, operation_ids)))
```

Comment:
it's a bit strange to make pool aware of the concept of node ID. It hurts the separation of concerns. But we can leave it like that for now